modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java

```diff
@@ -858,6 +1046,150 @@ private static ParameterMetadata createParameterMetadata(RelDataType parameterRo
+                CompletableFuture<PlanInfo> fut = ent.getValue();
+
+                if (currentCatalogVersion == key.catalogVersion() && isCompletedSuccessfully(fut)) {
+                    // no wait, already completed
+                    PlanInfo info = fut.join();
+
+                    if (info.sources.contains(tableId)) {
+                        ent.getKey().invalidate();
+                        statChanged = true;
+                    }
+                }
+            }
+
+            if (statChanged) {
+                recalculatePlans = true;
+            }
+        }
+
+        void start() {
+            planUpdater.scheduleAtFixedRate(() -> {
+                if (!recalculatePlans) {
+                    return;
+                }
+
+                CompletableFuture<PlanInfo> planFut = rePlanningFut.get();
+                if (planFut != null && !planFut.isDone()) {
+                    // some work still in progress
+                    return;
+                } else {
+                    rePlanningFut.set(null);
+                }
+
+                while (recalculatePlans) {
+                    recalculatePlans = false;
+
+                    for (Entry<CacheKey, CompletableFuture<PlanInfo>> ent : cache.entrySet()) {
+                        if (!ent.getKey().needInvalidate()) {
+                            continue;
+                        }
+
+                        CacheKey key = ent.getKey();
+                        CompletableFuture<PlanInfo> fut = cache.get(key);
+
+                        // can be evicted
+                        if (fut != null) {
+                            assert isCompletedSuccessfully(fut);
+
+                            PlanInfo info = fut.join();
+
+                            assert info.context != null && info.statement != null;
+
+                            int currentCatalogVersion = catalogVersionSupplier.getAsInt();
+
+                            if (currentCatalogVersion == key.catalogVersion()) {
+                                SqlQueryType queryType = info.statement.parsedResult().queryType();
+
+                                if (queryType != SqlQueryType.QUERY && queryType != SqlQueryType.DML) {
+                                    assert false : "Unexpected type: " + queryType;
+                                    continue;
+                                }
+
+                                PlanningContext planningContext = PlanningContext.builder()
+                                        .frameworkConfig(info.context.config())
+                                        .query(info.statement.parsedResult().originalQuery())
+                                        .plannerTimeout(plannerTimeout)
+                                        .catalogVersion(info.context.catalogVersion())
+                                        .defaultSchemaName(info.context.schemaName())
+                                        .parameters(info.context.parameters())
```

Comment:
Behavior depends on whether assertion is on or off.
Here should be either `assert` or `if-continue`